# Variant Annotation and Filtering

## Somatic SNVs and Indels
Variant-level annotation, filtering, and flagging of variants with further filter flags occur in the `SomaticCombineChannel` and `SomaticAnnotateMaf` processes. The union of variants that pass the somatic scoring models intrinsic to the callers (`FILTER="PASS"` in the VCF files) are combined, giving precedence to MuTect2 for any site where both callers detected a variant. 

The functional effect of variants is predicted using [VEP](https://www.ensembl.org/vep) using [vcf2maf](https://github.com/mskcc/vcf2maf), which also converts from VCF into a tab-delimited MAF file. See notes on use of [preferred transcript isoforms](reference-resources.md#preferred-transcript-isoforms) and [VEP annotation outputs](https://useast.ensembl.org/info/docs/tools/vep/vep_formats.html#output).


The following columns are added to the final MAF file, in addition to those added during the VEP annotation:
- `Strelka2FILTER`: Indicates that Strelka2 detected the variant but did not classify it as a somatic variant.
- `gnomAD_FILTER`: Indicates that the variant was detected in the gnomAD workflow, but ultimately _not_ classified as a _germline_ variant. Note that this is not used in current filtering schema.
- `RepeatMasker` and `EncodeDacMapability`: The variant locus is in a repeat, low-mappability, or hard-to-sequence region. More details in the [reference file description](reference-resources.md#repeatmasker-and-mappability-blacklist).
- `PoN`: Panel-of-normals, the number of normal samples in which the variant was detected. [More details on the implementation for exomes](wes-panel-of-normals.md). Under development for genomes, currently applies exome PoN. 
- `Ref_Tri`: Trinucleotide context of SNVs, normalized to pyrimidine-to-purine transversions.
- gnomAD allele frequencies, [more details here](gnomad.md):
    - Exomes, columns named `non_cancer_*`: Allele counts (`AC`) and frequencies (`AF`) for the variant in the non-TCGA population and sub-populations of gnomAD.
    - Genomes, columns named (`AC_*`|`AF_*`): Allele counts and frequencies for the variant in populations and sub-populations of gnomAD.
- Raw allele counts: total and strand specific (`*_fwd` and `*_rev`) allele `count` and `depth` for tumor (`t_*`) and normal (`n_*`). These are unfiltered values, in contrast to those from the variant callers, generated by [GetBaseCountsMultiSample](https://github.com/zengzheng123/GetBaseCountsMultiSample).
- `alt_bias`: For variants with a _raw_ depth of at least 5 reads, this is true if all _raw_ variant-supporting reads are on either the forward or reverse strand.
- `ref_bias`: For variants with a _raw_ depth of at least 5 reads, this is true if all _raw_ reads are on either the forward or reverse strand.
- Mutation hotspots, also see :
    - `snv_hotspot`: SNV hotspots.
    - `threeD_hotspot`: 3D hotspots, in contrast to abovementioned "linear" hotspots.
    - `indel_hotspot` and `indel_hotspot_type`: In-frame indel indel hotspots, and indication of whether it overlaps a prior indel hotspot locus (`prior`) or overlaps an SNV hotspot (`novel`).
    - `Hotspot`: `TRUE` if either a linear SNV or indel hotspot.
- OncoKB annotation:
    - `mutation_effect` and `oncogenic`: Indicate the functional effect of the mutation and whether it is deemed to be oncogenic
    - `LEVEL_*` and `Highest_level`: Indicates whether there is any drug at the given level of actionability and which is the highest level of actionability, if any. Note that this is cancer-type agnostic in current implementation.
    - `citations`: References for OncoKB annotation.
- Variant caller metadata (development feature, subject to changes):
    - MuTect2<sup>1</sup>:
        - `MBQ`: Median base quality, comma-separated for reference and alternate allele.
        - `MFRL`: Median fragment length, comma-separated for reference and alternate allele.
        - `MMQ`: Median mapping quality, comma-separated for reference and alternate allele.
        - `MPOS`: Median distance of variant from end of read.
        - `OCM`: Number of reads whose original alignment does not match the reference.
        - `RPA`: If tandem repeat, number of times repeated (can be comma-separated for reference and alternate allele).
        - `STR`: Boolean, indicating that variant is a short tandem repeat.
        - `ECNT`: Number of events in haplotype.
    - Strelka2<sup>2</sup>:
        - `MQ`: Root mean square mapping quality.
        - `SNVSB`: Strand bias for somatic SNVs.
        - `FDP`: Number of basecalls filtered from original read depth for tier 1* read counts, for tumor (`t_FDP`) and normal (`n_FDP`).
        - `SUBDP`: Number of reads below tier 1 mapping-quality threshold aligned across site, for tumor (`t_SUBDP`) and normal (`n_SUBDP`).
        - `RU`: If indel, smallest repeating sequence unit in inserted or deleted sequence.
        - `IC`: If indel, number of times `RU` is repeated in variant.

The `FILTER` column in the unfiltered MAF file, can contain any semi colon-separated combination of the following filter flags, or say `PASS`:
- `part_of_mnv`: The variant is likely part of another called multi-nucleotide variant (MNV).
- `multiallelic2`: Multiallelic loci, likely artifact. For variants called by Strelka2. The `2` is added due the presence of `multiallelic` flag in the MuTect2 VCFs.
- `strand_bias`, variants likely artifactual due to strand bias:
    - For variants called by Mutect2, if all supporting reads come from one strand and there are a least 10 reads on both strands in either normal or tumor sample.
    - For variants called by Strelka2, if the total alternate read count is above 10 and all of these fall on either strand; or low mapping-quality variant suffering from bias in both supporting reads and total reads.
- `caller_conflict`: Variant was detected by both callers, but did not pass Strelka2's thresholds for somatic variant calling.
- The following read depth-based flags are parameterized according to the sequencing platform, see the `exome.config` and `genome.config` files.
    - `low_vaf`: Variant falls below lower threshold for tumor variant allele fraction (VAF).
    - `low_t_depth`: Variant falls below lower threshold for total depth in the tumor.
    - `low_t_alt_count`: Variant falls below lower threshold for reads supporting variant allele in tumor.
    - `low_n_depth`: Variant falls below lower threshold for total depth in normal.
    - `high_n_alt_count`: Variant exceeds upper threshold for reads supporting variant allele normal.
    - `mappability`/`repeatmasker`: Variant falls in blacklisted genomic region.
    - `high_gnomad_pop_af`: Variant exceeds upper threshold for allele fraction in gnomAD.
    - `PoN`: Variant exceeds upper threshold for count in panel of normals.
    - `low_mapping_quality`: For indels called by Strelka2, variant falls below lower mapping quality threshold. 

<small><sup>1</sup>See the MuTect2 documentation for more information: https://software.broadinstitute.org/gatk/documentation/article?id=11005</small>\
<small><sup>2</sup>See the Strelka2 documentation for more information: https://github.com/Illumina/strelka/blob/v2.9.x/docs/userGuide/README.md</small>

### Whitelisting
Mutational hotspots, where the value in `Hotspot` is `TRUE`, are retained in the filtered MAF file, if they:
- Are flagged with `low_vaf` but the tumor VAF is at least 0.02.
- Are flagged with `low_mapping_quality`, `low_t_depth`, or `strand_bias`.
_Note that combinations of above filter flags results in filtering of the variant._
